### ISSUE TITLE
Restore topics on reconnection

### ISSUE THREAD

**Issue by ThePooN** [d-fischer/twitch-pubsub-client#3]
## Feature Request

It seems subscribed topics are currently not restored after reconnecting to Twitch PubSub. This means that currently, after a reconnection has occured, we're not receiving any event anymore.

**Comment by d-fischer** [comment]
The `PubSubClient` class is meant to be fairly low level and I don't intend on shipping auto-resubscription for it for now.
The higher-level `SingleUserPubSubClient`, however, should really resubscribe when reconnecting. I'll fix this on master before 2.0 will be started.

**Comment by ThePooN** [comment]
I think it should be implemented in `PubSubClient`. Imagine you want more higher-level clients such as `SingleUserPubSubClient` in the future, if you want to track events from multiple channels for example (that's my use case actually).  
It doesn't make sense having to implement it in every children classes when you can just implement it in the parent one.

**Comment by d-fischer** [comment]
The problem is that `PubSubClient` does not keep track about which topics have been subscribed to at all, so it can not resubscribe to them either. It would be necessary to keep track of them in *both* classes to allow both resubscription in `PubSubClient` as well as the high level of developer friendliness offered in `SingleUserPubSubClient`. This *will* at some point result in desync bugs and all sorts of weird errors nobody will understand.

**Comment by d-fischer** [comment]
For an alternative way to go about this, the `SingleUserPubSubClient` takes an optional argument where you can supply a `PubSubClient` instance to use. You can create multiple instances of `SingleUserPubSubClient` on top of the same `PubSubClient`.

**Comment by ThePooN** [comment]
I still disagree, you would just have to list the registered topics in `PubSubClient` and resubscribe to them at reconnection. It's not more complicated than that, it doesn't change anything on `SingleUserPubSubClient` neither.  
I don't possibly see how tracking the topics for automatic reconnection `PubSubClient` would trigger a "desync bug", if the current implementation of `SingleUserPubSubClient` works fine.

**Comment by ThePooN** [comment]
> For an alternative way to go about this, the `SingleUserPubSubClient` takes an optional argument where you can supply a `PubSubClient` instance to use. You can create multiple instances of `SingleUserPubSubClient` on top of the same `PubSubClient`.

What if multiple instances of `SingleUserPubSubClient` were registered for the same users?

**Comment by d-fischer** [comment]
> What if multiple instances of `SingleUserPubSubClient` were registered for the same users?

Assuming you mean multiple instances for the same user on top of the same `PubSubClient`:

As long as you're not listening to the same topics on more than one `SingleUserPubSubClient` instances, it should be fine. But that's something out of control of the library and also indicates that you should probably centralize the access to the clients in your app instead of creating more than one for the same user.

**Comment by d-fischer** [comment]
Quick example I threw together in 5 minutes (and I could probably ship this code in the library, now that I think about it):

```ts
class PubSubRegistry {
  private _rootClient = new PubSubClient();
  private _userClients = new Map<string, SingleUserPubSubClient>();

  getUserListener(userId) {
    if (!this._userClients.has(userId)) {
      const newClient = new SingleUserPubSubClient(twitchClient, this._rootClient);
      this._userClients.set(userId, newClient);
      return newClient;
    }
    return this._userClients.get(userId)!;
  }
}
```

**Comment by ThePooN** [comment]
Tbh this still feels over-complicated. Tracking the topics in `PubSubClient` doesn't sound out-of-scope of the class to me, while keeping the simplification stuff in `SingleUserPubSubClient`. It shouldn't be about making the `PubSubClient` class as bare as possible IMO, but to implement what would be commonly needed instead.

Also, I assumed `SingleUserPubSubClient` was extended from `PubSubClient` earlier but I realized I'm wrong, however I still think topics should be tracked in `PubSubClient` instead.

**Comment by d-fischer** [comment]
After some additional consideration, this was implemented in the above commit on the `next` branch and released as version `2.0.0-pre.2` as part of `BasicPubSubClient` (which is the old `PubSubClient`, but was renamed).

**Comment by ThePooN** [comment]
thanks for implementing this!
